### PR TITLE
Use an enum for Assets

### PR DESCRIPTION
**Pull request by D4nte**
- New Asset enum
- Refactor static rate configuration to make simpler
- Starting the work to have a better interface

### CODE REVIEW

File: src/actionSelector.ts [hunk 1]
```diff
@@ -71,40 +76,33 @@ export class ActionSelector {
...
-            log(
-              `Rate is not configured to buy ${alphaLedger}:${alphaAsset} & sell ${betaLedger}:${betaAsset}`
-            );
+          const tradeAmounts = {
+            buyAsset: alphaAsset,
+            sellAsset: betaAsset,
+            buyNominalAmount: alphaNominalAmount,
+            sellNominalAmount: betaNominalAmount
+          };
+          if (isTradeAcceptable(tradeAmounts, this.config.rates)) {
```
[the remark this answers is not in the document]
Comment (D4nte) [reply src/actionSelector.ts]: It is checked on line 57

File: src/asset.ts [hunk 2]
```diff
@@ -0,0 +1,15 @@
+export enum Asset {
+  Bitcoin = "bitcoin",
+  Ether = "ether"
+}
+
+export function toAsset(asset: string) {
```
[the remark this answers is not in the document]
Comment (D4nte) [reply src/asset.ts]: I'll look into it but tsc does not like your snippet of code:

image

[the remark this answers is not in the document]
Comment (bonomat) [reply src/asset.ts]: Your mistake lies in the Asset enum. 
```
    enum Asset{
      bitcoin,
      ethereum
    }
```

vs

```
    enum Asset{
      Bitcoin = "bitcoin",
      Ethereum = "ethereum"
    }
```

[the remark this answers is not in the document]
Comment (da-kami) [reply src/asset.ts]: Unfortunately this cannot be achieved with `keyof typeof Asset` because it would require a typecast using `<...>`, which tsc does not like. One can indeed achieve this behaviour (without linter & Co complaints...) by doing this: 
```
const alphaAsset: Asset = (Asset as any)[swap.properties.parameters.alpha_asset.name];
const betaAsset: Asset = (Asset as any)[swap.properties.parameters.beta_asset.name];
```
However, I find this confusing and not readable. One does not understand what is going on there without looking this up, which is timely. 

Thus, I vote for leaving the explicit conversion function in the code. It is way easier to understand that way. If we overlooked another possibility to achieve this please let me know :) 

Details: https://stackoverflow.com/questions/17380845/how-do-i-convert-a-string-to-enum-in-typescript

[the remark this answers is not in the document]
Comment (bonomat) [reply src/asset.ts]: Sure, it was just a recommendation. Did not want to escalate this :D

[the remark this answers is not in the document]
Comment (D4nte) [reply src/asset.ts]: The issue is that the `(Asset as any)` actually did not work well either :)

**Comment by D4nte** [comment]
@da-kami let's squash the WIP commit before merging :)

File: tests/configs/default.toml [hunk 3]
```diff
@@ -2,15 +2,13 @@ comitNodeUrl = "http://localhost:8000"
 seedWords = "abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon"
 maxRetries = 20
 
+# rates.buyAsset.sellAsset, buy = 1, sell is the configured maxSell
```
Comment (luckysori) [inline tests/configs/default.toml]: :wrench: This comment is very hard to understand. I would remove it.

File: tests/configs/default.toml [hunk 4]
```diff
@@ -2,15 +2,13 @@ comitNodeUrl = "http://localhost:8000"
...
-# Aiming for a "centralised exchange" feeling
-# ETH-BTC (ether.bitcoin) Rate is the number of Bitcoins for 1 Ether (so maths are BTC divided by ETH)
-# sell = Sell Ether
-# buy = Buy Ether
-# Cannot have [ether.bitcoin] and [bitcoin.ether] in same config file
-# These are all nominal units (Bitcoin, Ether, *not* Satoshi/Wei)
-sell = 0.0095
-buy = 0.0105
+maxSell = 0.0105
+# If Bob buys Bitcoin for Ether, for 1 Bitcoin, he wants to sell at most 105.26 Ether
```
Comment (luckysori) [inline tests/configs/default.toml]: :+1: I like these examples much better!

File: src/ratesConfig.ts [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+  buyAsset: Asset;
+  sellAsset: Asset;
+  buyNominalAmount: Big;
+  sellNominalAmount: Big;
+}
+
+export type RatesConfig = {
+  [buyAsset in Asset]: {
+    [sellAsset in Asset]?: {
+      maxSell: number;
```
Comment (luckysori) [inline src/ratesConfig.ts]: :wrench: This name is confusing. Isn't this just a `rate`?

File: src/ratesConfig.ts [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+    }
+  }
+};
+
+/// Returns Buy divided by Sell (Sell-Beta in exchange terms) rate based on the configuration
+export function isTradeAcceptable(
```
Comment (luckysori) [inline src/ratesConfig.ts]: :upside_down_face: I find the logic here hard to follow :confused: Might just be the names chosen.

**Comment by luckysori** [comment]
I have removed the `automagically-mergify-this` label so that you can look at my non-blocking comments! Feel free to merge without changes though :smile:

File: src/actionSelector.ts [hunk 7]
```diff
@@ -54,58 +56,55 @@ export class ActionSelector {
...
-      const betaLedger = swap.properties.parameters.beta_ledger.name;
-      const alphaAsset = swap.properties.parameters.alpha_asset.name;
-      const betaAsset = swap.properties.parameters.beta_asset.name;
+      const alphaLedger = toLedger(
+        swap.properties.parameters.alpha_ledger.name
+      );
+      const betaLedger = toLedger(swap.properties.parameters.beta_ledger.name);
+      const alphaAsset = toAsset(swap.properties.parameters.alpha_asset.name);
+      const betaAsset = toAsset(swap.properties.parameters.beta_asset.name);
+      if (!alphaAsset || !betaAsset || !alphaLedger || !betaLedger) {
```
[the remark this answers is not in the document]
Comment (D4nte) [reply src/actionSelector.ts]: Yes we considered it but I don't think it's worth the extra code. The user should easily recognise which asset isn't setup/supported.

File: src/config.ts [hunk 8]
```diff
@@ -83,15 +83,24 @@ export class Config {
       : uriPath;
   }
 
-  public isSupported(ledgerName: string, assetName: string) {
-    switch (ledgerName + assetName) {
-      case "bitcoinbitcoin":
-        return typeof this.bitcoinConfig === "object";
-      case "ethereumether":
-        return typeof this.ethereumConfig === "object";
+  public isValidConfiguration(ledger: Ledger): boolean {
```
[the remark this answers is not in the document]
Comment (D4nte) [reply src/config.ts]: This name is more explicit to say whether there is a valid configuration for this ledger. Bobtimus could support a ledger but the user could not configure it. The `isSupport` in this case does not sound right.

File: src/actionSelector.ts [hunk 9]
```diff
@@ -54,58 +56,55 @@ export class ActionSelector {
...
+      }
+
+      const alphaNominalAmount = toNominalUnit(
+        swap.properties.parameters.alpha_asset
+      );
+      const betaNominalAmount = toNominalUnit(
+        swap.properties.parameters.beta_asset
+      );
+
+      if (!alphaNominalAmount || !betaNominalAmount) {
```
[the remark this answers is not in the document]
Comment (D4nte) [reply src/actionSelector.ts]: The logs prints both so you'd see which one is undefined, not too keen to have an array of `ifs`.

File: src/actionSelector.ts [hunk 10]
```diff
@@ -1,9 +1,11 @@
 import Big from "big.js";
 import debug from "debug";
 import { Action, Entity } from "../gen/siren";
+import { toAsset } from "./asset";
 import { Swap, toNominalUnit } from "./comitNode";
 import { Config } from "./config";
-import { getBuyDivBySellRate, isProfitable } from "./rates";
+import { toLedger } from "./ledger";
+import { isTradeAcceptable } from "./ratesConfig";
```
[the remark this answers is not in the document]
Comment (D4nte) [reply src/actionSelector.ts]: I'll split up.

File: src/config.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (D4nte) [reply src/config.ts]: I renamed

File: src/ratesConfig.ts (hunk 5 above)
Comment (D4nte) [reply src/ratesConfig.ts]: We'll rename everything in the next PR that is rebased on this.

File: src/ratesConfig.ts (hunk 6 above)
Comment (D4nte) [reply src/ratesConfig.ts]: It's maths, you know the maths, maybe proposed new names? Happy to call and explain. I have attempted a rename.

**Comment by D4nte** [comment]
> Bear in mind that none of my comments are blocking, but I did find some of the code I looked at hard to understand. I don't know how much we care about code clarity in Bobtimus though.

The main question is: is it harder or easier to understand from before?
This is hard anyway, so it's not like we made something easy to conceive into hard code. It's a complex problem and we have less-complex code now.

I welcome suggestions that'd make the code easier to reason about....

File: src/ratesConfig.ts [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+  const sellAmount = tradeAmounts.sellNominalAmount;
+  const rate = ratesConfig[buyAsset][sellAsset];
+
+  if (!rate || !rate.maxSell) {
+    log(`Rate not configured for buy: ${buyAsset}, sell: ${sellAsset}`);
+    return false;
+  }
+  const maxSellForAmountOne = rate.maxSell;
+
+  const maxSell = buyAmount.mul(maxSellForAmountOne);
```
Comment (thomaseizinger) [inline src/ratesConfig.ts]: :question: Is `maxSell` a factor?

Comment (D4nte) [reply src/ratesConfig.ts]: Refer to:
```
# If Bob buys Bitcoin for Ether, for 1 Bitcoin, he wants to sell at most 105.26 Ether
[rates.bitcoin.ether]
maxSell = 105.26
```

**Comment by D4nte** [comment]
> I'll trust you that the logic is on point. The wording could maybe be improved :)

Please provide suggestions on how it can be improved. This is the best we could come up with @da-kami.

File: src/ratesConfig.ts (hunk 5 above)
Comment (D4nte) [reply src/ratesConfig.ts]: I did some rename but more rename will come with the following PR that is already WIP

File: src/ratesConfig.ts (hunk 6 above)
Comment (D4nte) [reply src/ratesConfig.ts]: I had a chat with @thomaseizinger and we did not come with something better. Maybe let's chat face-to-face :)
